public/components/homepage.js

```diff
@@ -0,0 +1,182 @@
+    if (!e.keyCode || e.keyCode == 13) {
+      let searchWarning = document.getElementById('search-warning');
+      let searchName = document.getElementById('search-name').value;
+      if (searchName.length < 3) {
+        searchWarning.innerHTML = "At least 3 characters needed.";
+        return;
+      }
+      if (searchName.match(/[^A-Za-z ]/)) {
+        searchWarning.innerHTML = "Only alphabetic characters allowed.";
+        return;
+      }
+      searchWarning.innerHTML = "";
+      let category = document.getElementById('filter-category-dropdown').value;
+      let location = document.getElementById('filter-location-dropdown').value;
+      let query = `/search?keyword=${searchName}`;
+        // + (category ? `&category=${category}` : '')
+        // + (location ? `&location=${location}` : '');
+      ajaxRequest('GET', query, { ajax : true }, {}, renderFreelancers);
+    }
+  }
+  render() {
+    return (
+      <div id="search-container">
+        <div id="search-bar">
+          <input id="search-name" placeholder="What service do you need?" onKeyDown={this.searchRequest} />
+          <button id="search-btn" onClick={this.searchRequest}>Search</button>
+        </div>
+        <div id="search-warning"></div>
+      </div>
+    );
+  }
+}
+
+// Container for all filters
+class FiltersContainer extends React.Component {
+  render() {
+    let categories = [];
+    for (let i = 0; i < this.props.categories.length; ++i) {
+      let category = this.props.categories[i].categoryName;
+      categories.push(<option value={category} key={i}>{category}</option>);
+    }
+    let locations = [];
+    for (let i = 0; i < this.props.locations.length; ++i) {
+      let location = this.props.locations[i]
+      locations.push(<option value={location} key={i}>{location}</option>);
+    }
+    return (<div id="filters-container">
+      <div id="filters">
+        <div id="filter-category">
+          <span>Category: </span>
+          <select id="filter-category-dropdown" defaultValue="">
+            <option value="">Anything</option>
+            {categories}
+          </select>
+        </div>
+        <div id="filter-location">
+          <span>Location: </span>
+          <select id="filter-location-dropdown" defaultValue="">
+            <option value="">Anywhere</option>
+            {locations}
+          </select>
+        </div>
+      </div>
+    </div>);
+  }
+}
+
+// Button that redirects to `add freelancer form`
+class FreelancerCreateBtn extends React.Component {
+  redirectFreelancerForm() {
+    document.location = '/freelance/new';
+  }
+  render () {
+    return (<button id="freelancer-create-btn" onClick={this.redirectFreelancerForm}>Add a new freelancer</button>);
+  }
+}
+
+// Card which displays a freelancer's information
+class FreelancerCard extends React.Component {
+  redirectFreelancer(freelancer) {
+    return function() {
+      document.location = `/freelance/${freelancer.props._id}`;
```

Comment:
There's a property array `links` in each freelancer that contains the links. The link to the freelancer itself (`"rel" : "self"`) can be found at `freelancer.links[0].href`, and it's the full link to the freelancer.
If we use that, if we change the routing then the change will be automatic for the client side.